### PR TITLE
e2e: formatting enhancements

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Currently some objects get rendered in all of their gory details by Gomega, which makes them unreadable and forces Gomega to truncate before coming to the interesting parts:
```
    Aug 17 09:55:26.691: INFO: Unexpected error: start pod with inline resource claim: 
        <*pod.timeoutError | 0xc004274e40>: {
            msg: "timed out while waiting for pod dra-5838/tester-1 to be running",
            observedObjects: [
                <*v1.Pod | 0xc002123800>{
                    TypeMeta: {Kind: "", APIVersion: ""},
                    ObjectMeta: {
                        Name: "tester-1",
                        GenerateName: "",
                        Namespace: "dra-5838",
                        SelfLink: "",
                        UID: "c3a4dfa5-5297-44e0-acdc-bdfb82035b80",
                        ResourceVersion: "379",
                        Generation: 0,
                        CreationTimestamp: {
                            Time: {
                                wall: 0,
                                ext: 63796319426,
                                loc: {
                                    name: "Local",
...

                                        {w...

    Gomega truncated this representation as it exceeds 'format.MaxLength'.
    Consider having the object provide a custom 'GomegaStringer' representation
    or adjust the parameters in Gomega's 'format' package.

    Learn more here: https://onsi.github.io/gomega/#adjusting-output
```

By teaching Gomega to use YAML, we get a nicer, familiar and more compact representation of objects or structs used inside such objects. YAML is used as soon as a struct has JSON tags for all of its fields, which implies that it can be represented completely by JSON or YAML.

Even with the more compact representation, we still need to deal with long messages. Therefore this PR also:
- makes the truncation limit configurable
- automatically truncates failure messages, with full details in the output

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/111884

#### Special notes for your reviewer:

Builds on top of https://github.com/kubernetes/kubernetes/pull/113367

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: test/e2e/framework/test_context.go
Comment (pohly) [inline test/e2e/framework/test_context.go:536]: /hold

Let's first discuss in https://github.com/kubernetes/test-infra/pull/27831#discussion_r1008446740 what a good maximum size would be.

Comment (spiffxp) [reply test/e2e/framework/test_context.go:536]: [Here's what we use in triage](https://github.com/kubernetes/test-infra/blob/2c20bdb10a0e3177a1163cbb2b4f83b450d7df70/triage/summarize/summarize.go#L31-L32)

10K when clustering things together, 100K when displaying failure text (IIRC this came about from integration tests having a ton of output)

So I'm fine with at least 10K, I feel like this is something we can adjust as followup if we don't like it

**Review by spiffxp — APPROVED** [review]
/triage accepted
/priority important-soon
/approve
/lgtm

**Comment by spiffxp** [comment]
Remove /hold if you want to merge as is, or I'll trust someone else to lgtm when you change the max failure message size

File: test/utils/format/format_test.go
Comment (aojea) [inline test/utils/format/format_test.go:66]: what is this doing?

Comment (pohly) [reply test/utils/format/format_test.go:66]: It removes the hex address of objects from the output before comparing it against the expected text. Otherwise the output would be different for each test run.

File: test/e2e/framework/test_context.go
Comment (pohly) [reply test/e2e/framework/test_context.go:536]: That's significantly larger than what we produce here, so 5000 should be safe. Let's merge and see how well it works in practice.

/hold cancel

**Pull request merged**
